### PR TITLE
Add new build flag for local-run command

### PR DESCRIPTION
**Pull request by NikhilSharmaWe**
## Description

This PR adds a new flag `build` for `local-run` command. If the value of the `build` flag is set to `true`(default) then the function is built before `local-run`.

## Motivation and Context

- Fixes: #953

## How Has This Been Tested?

I have tested the new command locally and it works as expected.
- With `--build=false`, `local-run` command does not reflect the new changes made in the funtion.
- With `--build=true` it is not the case, the changes are reflected in the run, since `build` command logic is being executed before each `local-run`.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: commands/localBuild.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	newLocalBuildCmd.Flags().BoolVar(&shrinkwrap, "shrinkwrap", false, "Just write files to ./build/ folder for shrink-wrapping")
+	newLocalBuildCmd.Flags().StringArrayVarP(&buildArgs, "build-arg", "b", []string{}, "Add a build-arg for Docker (KEY=VALUE)")
+	newLocalBuildCmd.Flags().StringArrayVarP(&buildOptions, "build-option", "o", []string{}, "Set a build option, e.g. dev")
+	newLocalBuildCmd.Flags().Var(&tagFormat, "tag", "Override latest tag on function Docker image, accepts 'latest', 'sha', 'branch', or 'describe'")
+	newLocalBuildCmd.Flags().StringArrayVar(&buildLabels, "build-label", []string{}, "Add a label for Docker image (LABEL=VALUE)")
+	newLocalBuildCmd.Flags().StringArrayVar(&copyExtra, "copy-extra", []string{}, "Extra paths that will be copied into the function build context")
+	newLocalBuildCmd.Flags().BoolVar(&envsubst, "envsubst", true, "Substitute environment variables in stack.yml file")
+	newLocalBuildCmd.Flags().BoolVar(&quietBuild, "quiet", false, "Perform a quiet build, without showing output from Docker")
+	newLocalBuildCmd.Flags().BoolVar(&disableStackPull, "disable-stack-pull", false, "Disables the template configuration in the stack.yml")
```
Comment (NikhilSharmaWe) [inline commands/localBuild.go:33]: Can create functions for the flags and options for avoiding repetition in the `build.go`, `local_run.go` and `localBuild.go`, right?

**Comment by alexellis** [comment]
Hi, thanks for the PR.

I think what we're looking for is a `--build` (default of `true`) flag for the `faas-cli local-run` command.

Could you change the code and update the PR?

Alex

**Comment by alexellis** [comment]
We have a community call tomorrow if you can make it? https://docs.openfaas.com/community/

File: commands/local_run.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ type runOptions struct {
 	extraEnv map[string]string
 	output   io.Writer
 	err      io.Writer
+	build    bool
 }
```
Comment (NikhilSharmaWe) [inline commands/local_run.go:31]: Made changes
@alexellis Could you please review?

File: commands/local_run.go [hunk 3]
```diff
@@ -82,13 +90,28 @@ services deployed within your OpenFaaS cluster.`,
...
+	cmd.Flags().BoolVar(&opts.build, "build", true, "Build function prior to local-run")
 	cmd.Flags().IntVarP(&opts.port, "port", "p", 8080, "port to bind the function to")
 	cmd.Flags().StringVar(&opts.network, "network", "", "connect function to an existing network, use 'host' to access other process already running on localhost. When using this, '--port' is ignored, if you have port collisions, you may change the port using '-e port=NEW_PORT'")
 	cmd.Flags().StringToStringVarP(&opts.extraEnv, "env", "e", map[string]string{}, "additional environment variables (ENVVAR=VALUE), use this to experiment with different values for your function")
 
 	return cmd
 }
 
+func localBuild(cmd *cobra.Command, args []string) error {
+	err := preRunBuild(cmd, args)
```
Comment (alexellis) [inline commands/local_run.go:102]: I think I'd probably inline these with `if err:= .... ; err !=nil {`

How does this work when there are flags that the build command uses like `--build-arg` - which are not present in the local run command's flags?

I.e. `faas-cli build --build-arg GO111MODULE=on`

Have a look at how `faas-cli up` works

Comment (NikhilSharmaWe) [reply commands/local_run.go]: Made changes

**Comment by alexellis** [comment]
That looks more like what I was expecting to see :+1: 

@NikhilSharmaWe we'll review this / try it out on the community call in 20 minutes from now.

**Review by LucasRoesler — APPROVED** [review]
@alexellis i just tested this locally and everything works 100% as expected. the implementation also looks good

**Review by alexellis — APPROVED** [review]
LGTM

Thanks for working through this with us :+1:

**Comment by alexellis** [comment]
This'll be available soon via https://github.com/openfaas/faas-cli/releases/tag/0.16.4

`arkade get faas-cli` will get the latest version once the build is complete.

**Pull request merged**